crates/compiler/mono/src/decision_tree.rs

```diff
@@ -2168,6 +2171,69 @@ fn test_always_succeeds(test: &Test) -> bool {
     }
 }
 
+fn sort_edge_tests_by_priority(edges: &mut [Edge<'_>]) {
+    use std::cmp::{Ordering, Ordering::*};
+    use GuardedTest::*;
+    edges.sort_by(|(t1, _), (t2, _)| match (t1, t2) {
+        // Guarded takes priority
+        (GuardedNoTest { .. }, GuardedNoTest { .. }) => Equal,
+        (GuardedNoTest { .. }, TestNotGuarded { .. }) | (GuardedNoTest { .. }, Placeholder) => Less,
+        // Interesting case: what test do we pick?
+        (TestNotGuarded { test: t1 }, TestNotGuarded { test: t2 }) => order_tests(t1, t2),
+        // Otherwise we are between guarded and fall-backs
+        (TestNotGuarded { .. }, GuardedNoTest { .. }) => Greater,
+        (TestNotGuarded { .. }, Placeholder) => Less,
+        // Placeholder is always last
+        (Placeholder, Placeholder) => Equal,
+        (Placeholder, GuardedNoTest { .. }) | (Placeholder, TestNotGuarded { .. }) => Greater,
+    });
+
+    fn order_tests(t1: &Test, t2: &Test) -> Ordering {
+        match (t1, t2) {
+            (
+                Test::IsListLen {
+                    bound: bound_l,
+                    len: l,
+                },
+                Test::IsListLen {
+                    bound: bound_m,
+                    len: m,
+                },
+            ) => {
+                // List tests can either check for
+                //   - exact length (= l)
+                //   - a size greater or equal to a given length (>= l)
+                // (>= l) tests can be superset of other tests
+                //   - (>= m) where m > l
+                //   - (= m)
+                // So, if m > l, we enforce the following order for list tests
+                //   (>= m) then (= l) then (>= l)
+                match m.cmp(l) {
+                    Less => Less, // (>= m) then (>= l)
+                    Greater => Greater,
+
+                    Equal => {
+                        use ListLenBound::*;
+                        match (bound_l, bound_m) {
+                            (Exact, AtLeast) => Less, // (= l) then (>= l)
+                            (AtLeast, Exact) => Greater,
+
+                            (AtLeast, AtLeast) | (Exact, Exact) => Equal,
+                        }
+                    }
+                }
```

Comment:
I think https://doc.rust-lang.org/std/cmp/enum.Ordering.html#method.then would be nice here